### DISCUSSION TITLE
Unable to read Statepoint File

### DISCUSSION THREAD

**Discussion by AnthonyB08** [neams-th-coe/cardinal#618] [Q&A]
Hi all,

I managed to run Cardinal, with open mc and Moose, with post-processing primarily accomplished by Moose. However, for me to fully understand the built-in normalization of Cardinal I am trying to open the StatePoint file with python API 
(with openmc.StatePoint("statepoint.25.h5") as sp:) and I am returned with this error: 

statepoint file has a version of 17.0 which is not consistent with the version expected by OpenMC, 18

What does the community propose I do to be able to read-in the statepoint file. 

-Thanks,
Anthony Bowers

**Comment by aprilnovak** [comment]
Hmm, that's strange. Can you try running the OpenMC case in "standalone" mode to see if the same issue persists? We're not doing any calling/writing of statepoint files ourselves (e.g. OpenMC is handling all that).

```
cd <your case directory>
openmc
<try loading the statepoint file>
```

**Reply by aprilnovak** [reply]
@AnthonyB08 by the way, would you like me to add you to our Cardinal slack channel? We may be able to more quickly iterate through questions like this by hooking up INL folks in those conversations - really appreciate the software stress testing :) If so, please send me your email to anovak at anl.gov

**Reply by AnthonyB08** [reply]
Running the standalone worked, but running in jupyter-notebook with " mpiexec -np 2 ../../cardinal-opt -i fluxV3.i --n-threads=2" did not allow me to access the state point file, very strange.

Sure, Im never used slack before but I can surely get the hang of it with time. I will download and create an account now.

**Reply by aprilnovak** [reply]
You'll need to send me your email so that I can add you to the slack space - I'm not sure if you'll be able to request an account yourself.

**Reply by aprilnovak** [reply]
@lewisgross1296 you frequently read the statepoint files from Cardinal into separate OpenMC postprocessing python scripts, right? Have you ever encountered this?

**Reply by AnthonyB08** [reply]
I have sent the email. 

I wonder if this error has anything to do with DAGMC. Openmc standalone is working, but when I do open mc standalone with a DAGMC geometry XML I get the error "Openmc is not configured with DAGMC enabled", which is expected. However, if I run through cardinal-opt I am able to simulate with DAGMC geometry file but am unable to open the statepoint file.

**Reply by lewisgross1296** [reply]
Do you have your post-processing script on github? I'd like to see more than just the ` openmc.StatePoint("statepoint.25.h5") as sp` line since this is correct OpenMC syntax. It may be an issue with your install/dependencies. It also may be something with Jupyter since I typically run post processing scripts from the terminal with python 3.10.5 (these days)

**Reply by AnthonyB08** [reply]
@lewisgross1296  Sure, give me some time to set up a repository and I will make it public and link it here. Thank you!

**Reply by lewisgross1296** [reply]
Though I typically do something like this
```
import openmc
#open statepoint
sp = openmc.StatePoint('statepoint.25.h5')
```
Not sure it is intended to be used via `with`, but not sure that is wrong per se.

**Reply by AnthonyB08** [reply]
Here is the repository. This is a verification analysis on DAGMC and traditional OpenMC Geometries 
https://github.com/AnthonyB08/DAGMC_MOOSE

**Reply by lewisgross1296** [reply]
Do you get the same error when you replace `with openmc.StatePoint("statepoint.20.h5") as sp:` with the line I wrote above? If so I think this is maybe something to do with your OpenMC install/executable.

**Reply by AnthonyB08** [reply]
Yes. How did you path your openmc in the bashrc for importing the library in python?

**Reply by AnthonyB08** [reply]
I have determined the issue. I had openmc source installation in my python3.10 site packages (/.local) and the statepoint version was 18. In the cardinal/contrib/openmc/openmc  the statepoint version is 17

**Reply by AnthonyB08** [reply]
I need to figure out how to path the openmc ( in cardinal) for jupyter.

**Reply by lewisgross1296** [reply]
I'm not sure this is an environment variable issue but this is how my environment variables are set up for Cardinal and OpenMC
```
# Cardinal Environment 
# build without NEK
export ENABLE_NEK=false

# [REQUIRED] you must set the location of the root directory of the NekRS install;
# this will be the 'install' directory at the top level of the Cardinal repository.
# Be sure to adjust this to wherever you have cloned the Cardinal repository
export NEKRS_HOME=$HOME/cardinal/install

# [REQUIRED IF USING MOOSE CONDA ENV] you must set the location of the HYPRE
# libraries (so that a libHYPRE.so or something similar is located at
# HYPRE_DIR/lib). This is needed because NekRS uses the same HYPRE built with MOOSE
# export HYPRE_DIR=$CONDA_PREFIX

# [OPTIONAL] it's a good idea to explicitly note that you are using MPI compiler wrappers
export CC=mpicc
export CXX=mpicxx
export FC=mpif90

# [OPTIONAL] if running with OpenMC, you will need cross section data at runtime;
# you will need to set this variable to point to a 'cross_sections.xml' file.
# Be sure to adjust this to wherever you have cross section data.
export OPENMC_CROSS_SECTIONS=${HOME}/cross_sections/endfb71_hdf5/cross_sections.xml

# without setting HDF5_ROOT, the default is PETSc's HDF5, which is fine

# using multiple options causes multiple libraries to be built with different settings
export METHODS='opt oprof debug'

# current method to build
export METHOD='opt'

# set the OCCA Backend
export NEKRS_OCCA_MODE_DEFAULT=CPU

# set the PYTHONPATH
export PYTHONPATH=/home/lgross/cardinal/contrib/moose/python:$PYTHONPATH

export MOOSE_DIR=/home/lgross/cardinal/contrib/moose

# openmc
export OPENMC_ENDF_DATA=/home/lgross/endf-b-vii.1
# set the OPENMC_DIR when testing new features
export OPENMC_DIR=/home/lgross/openmc
# add openmc executable to path
export PATH=/home/lgross/openmc/build/bin:$PATH
```

This feels like more of an OpenMC issue than a Cardinal one. I just looked for this issue on the [OpenMC discourse](https://openmc.discourse.group/) and it seems like you are not the only one with this problem. Maybe you could comment on [this thread](https://openmc.discourse.group/t/problem-to-open-output-files-of-older-openmc-version/2582/2) to see if someone at OpenMC has more helpful tips

**Reply by lewisgross1296** [reply]
Actually [this thread](https://openmc.discourse.group/t/oserror-statepoint-file-has-a-version-of-17-0-which-is-not-consistent-with-the-version-expected-by-openmc-18/2304) seems to be the exact same error as you and Patrick gave an answer as what to do.

**Reply by aprilnovak** [reply]
@AnthonyB08 when was the last time you built Cardinal? Last week, I update the OpenMC submodule (it was previously pointing to a version from July, so there was a big gap there). Maybe if you do the following the version issue will be fixed?

```
cd cardinal
git pull upstream devel
./scripts/get-dependencies.sh
make -j8
```

**Reply by AnthonyB08** [reply]
@lewisgross1296  I am curious then, where is your sys path for import openmc? 

@aprilnovak  I am trying that now, will update ASAP.

**Reply by lewisgross1296** [reply]
Maybe this is what you were wondering?
```
lgross@ulam:~/openmc (s2patch) $ python
Python 3.10.5 (main, Jun 30 2022, 12:02:41) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> print(sys.path)
['', '/home/lgross/cardinal/contrib/moose/python', '/home/lgross/openmc', '/home/lgross/.pyenv/versions/3.10.5/lib/python310.zip', '/home/lgross/.pyenv/versions/3.10.5/lib/python3.10', '/home/lgross/.pyenv/versions/3.10.5/lib/python3.10/lib-dynload', '/home/lgross/.pyenv/versions/3.10.5/lib/python3.10/site-packages']
```
Could be worth looking into pyenv for managing OpenMC's python API. April suggested this to me a while back.
